### PR TITLE
[Web] Update to eslint 9

### PR DESCRIPTION
**Pull request by Faless**
Draft since it needs more work to match the old (stricter) rules.

### CODE REVIEW

**Comment by patwork** [comment]
Note, in flat config mode for eslint 9+ you can create one shared config like this (not finished example):

```javascript
import globals from "globals";
import pluginJs from "@eslint/js";
import pluginReference from "eslint-plugin-html";

export default [
	pluginJs.configs.recommended,
	// pluginJs.configs.all,

	{
		rules: {
			"indent": ["error", "tab"],
			"curly": ["error", "all"],
			"quote-props": ["error", "consistent"],
			"no-self-assign": "off",
			"no-unused-vars": "warn", // FIXME
		},
	},

	{
		files: ["js/engine/**/*.js"],
		languageOptions: {
			globals: {
				...globals.browser,
				"Features": true,
				"Godot": true,
				"InternalConfig": true,
				"Preloader": true,
			},
		},
	},

	{
		files: ["js/jsdoc2rst/**/*.js"],
		languageOptions: {
			globals: globals.node,
		},
	},

	{
		files: ["js/libs/**/*.js"],
		languageOptions: {
			globals: {
				...globals.browser,
				"autoAddDeps": true,
				"Browser": true,
				"ERRNO_CODES": true,
				"FS": true,
				"GL": true,
				"GodotConfig": true,
				"GodotEventListeners": true,
				"GodotFS": true,
				"GodotOS": true,
				"GodotRuntime": true,
				"HEAP32": true,
				"HEAP8": true,
				"HEAPF32": true,
				"HEAPU8": true,
				"IDBFS": true,
				"IDHandler": true,
				"LibraryManager": true,
				"mergeInto": true,
				"XRWebGLLayer": true,
			},
		},
	},

	{
		files: ["html/**/*.js"],
		languageOptions: {
			globals: {
				...globals.browser,
				"___GODOT_CACHE___": true,
				"___GODOT_ENSURE_CROSSORIGIN_ISOLATION_HEADERS___": true,
				"___GODOT_OPT_CACHE___": true,
				"onClientMessage": true,
			},
		},
	},

	{
		files: ["html/**/*.html"],
		plugins: {
			"eslint-plugin-html": pluginReference,
		},
		languageOptions: {
			globals: {
				...globals.browser,
				"Engine": true,
				"$GODOT_CONFIG": true,
				"$GODOT_PROJECT_NAME": true,
				"$GODOT_THREADS_ENABLED": true,
				"___GODOT_THREADS_ENABLED___": true,
			},
		},
	},

	{
		ignores: [
			"eslint.config.mjs",
			"**/*.externs.js",
		],
	},
];
```

**Comment by Faless** [comment]
@patwork amazing, thanks a lot :blue_heart: !

I think `no-unused-vars` should be:
```
"no-unused-vars": ["error", { "args": "none", "caughtErrors": "none" }]
```

**Comment by Faless** [comment]
As discussed with @patwork this needs some more work to match the old (stricter) rules.

I'm leaving this PR in draft as a base for future work.

File: platform/web/package.json [hunk 1]
```diff
@@ -6,13 +6,13 @@
...
-    "lint:modules": "eslint \"../../modules/**/*.js\" --no-eslintrc -c .eslintrc.libs.js",
-    "lint:tools": "eslint \"js/jsdoc2rst/**/*.js\" --no-eslintrc -c .eslintrc.engine.js",
-    "lint:html": "eslint \"../../misc/dist/html/*.html\" --no-eslintrc -c .eslintrc.html.js",
-    "format": "npm run format:engine && npm run format:libs && npm run format:modules && npm run format:tools && format:sw && npm run format:html",
+    "lint:engine": "cd ../../ && eslint \"platform/web/js/engine/*.js\" --no-config-lookup -c platform/web/eslint.config.mjs",
+    "lint:libs": "cd ../../ && eslint \"platform/web/js/libs/*.js\" --no-config-lookup -c platform/web/eslint.config.mjs",
+    "lint:tools": "cd ../../ && eslint \"platform/web/js/jsdoc2rst/**/*.js\" --no-config-lookup -c platform/web/eslint.config.mjs",
+    "lint:modules": "cd ../../ && eslint \"modules/**/*.js\" --no-config-lookup -c platform/web/eslint.config.mjs",
+    "lint:sw": "cd ../../ && eslint \"misc/dist/html/service-worker.js\" --no-config-lookup -c platform/web/eslint.config.mjs",
+    "lint:html": "cd ../../ && eslint \"misc/dist/html/*.html\" --no-config-lookup -c platform/web/eslint.config.mjs",
```
Comment (Faless) [inline platform/web/package.json:14]: We could probably now unify this as a single command I think:

```
cd ../../ && eslint \"platform/web/js/engine/*.js\" \"platform/web/js/libs/*.js\" \"platform/web/js/jsdoc2rst/**/*.js\" \"modules/**/*.js\" \"misc/dist/html/service-worker.js\" eslint \"misc/dist/html/*.html\" --no-config-lookup -c platform/web/eslint.config.mjs
```

Comment (Faless) [reply platform/web/package.json:14]: Or even:
```
cd ../../ && eslint \"platform/web/**/*.js\" \"modules/**/*.js\" \"misc/dist/html/*.js\" \"misc/dist/html/*.html\" --no-config-lookup -c platform/web/eslint.config.mjs
```

**Comment by patwork** [comment]
1) I was wondering what is "onClientMessage" in service worker? I could not find documentation for such a function anywhere.

2) eslint upgrade will also require the removal of "eslint-disable" directives in several javascript files

3) I was wondering whether, instead of ignoring errors in catch(), to add in each one something like console.warn(e)

4) thanks to flat config, scripts in package.json can be simplified:

```json
"scripts": {
  "docs": "jsdoc --template js/jsdoc2rst/ js/engine/engine.js js/engine/config.js js/engine/features.js --destination ''",
  "lint": "cd ../.. && eslint --no-config-lookup --config ./platform/web/eslint.config.mjs ./platform/web ./modules ./misc/dist/html",
  "fix": "cd ../.. && eslint --fix --no-config-lookup --config ./platform/web/eslint.config.mjs ./platform/web ./modules ./misc/dist/html",
},
```

**Comment by patwork** [comment]
All "eslint-disable" directives in current master:

```
./misc/dist/html/editor.html:function closeWelcomeModal(dontShowAgain) { // eslint-disable-line no-unused-vars
./misc/dist/html/editor.html:function clearPersistence() { // eslint-disable-line no-unused-vars
./misc/dist/html/editor.html:function closeEditor() { // eslint-disable-line no-unused-vars
./misc/dist/html/service-worker.js:							console.error('Network error: ', e); // eslint-disable-line no-console
./modules/webxr/native/library_godot_webxr.js:			runtimeKeepalivePush(); // eslint-disable-line no-undef
./modules/webxr/native/library_godot_webxr.js:				runtimeKeepalivePop(); // eslint-disable-line no-undef
./modules/webxr/native/library_godot_webxr.js:			const gl_context_handle = _emscripten_webgl_get_current_context(); // eslint-disable-line no-undef
./modules/webxr/native/library_godot_webxr.js:				GodotWebXR.gl_binding = new XRWebGLBinding(session, gl); // eslint-disable-line no-undef
./platform/web/js/engine/config.js:const EngineConfig = {}; // eslint-disable-line no-unused-vars
./platform/web/js/engine/config.js:const InternalConfig = function (initConfig) { // eslint-disable-line no-unused-vars
./platform/web/js/engine/config.js:			console.log.apply(console, Array.from(arguments)); // eslint-disable-line no-console
./platform/web/js/engine/config.js:			console.error.apply(console, Array.from(arguments)); // eslint-disable-line no-console
./platform/web/js/engine/engine.js:	function Engine(initConfig) { // eslint-disable-line no-shadow
./platform/web/js/engine/features.js:const Features = { // eslint-disable-line no-unused-vars
./platform/web/js/engine/preloader.js:const Preloader = /** @constructor */ function () { // eslint-disable-line no-unused-vars
./platform/web/js/jsdoc2rst/publish.js:/* eslint-disable strict */
./platform/web/js/jsdoc2rst/publish.js:				console.log(d); // eslint-disable-line no-console
./platform/web/js/jsdoc2rst/publish.js:				console.log(`Undefined symbol! ${d.memberof}`); // eslint-disable-line no-console
./platform/web/js/libs/library_godot_display.js:			const gl_context_handle = _emscripten_webgl_get_current_context(); // eslint-disable-line no-undef
./platform/web/js/libs/library_godot_display.js:		window.alert(GodotRuntime.parseString(p_text)); // eslint-disable-line no-alert
./platform/web/js/libs/library_godot_display.js:			alert('WebGL context lost, please reload the page'); // eslint-disable-line no-alert
./platform/web/js/libs/library_godot_javascript_singleton.js:				const global_eval = eval; // eslint-disable-line no-eval
./platform/web/js/libs/library_godot_javascript_singleton.js:				eval_ret = eval(js_code); // eslint-disable-line no-eval
./platform/web/js/libs/library_godot_os.js:			return Module['locateFile'](file); // eslint-disable-line no-undef
./platform/web/js/libs/library_godot_runtime.js:			return wasmTable.get(ptr); // eslint-disable-line no-undef
./platform/web/js/libs/library_godot_runtime.js:			err.apply(null, Array.from(arguments)); // eslint-disable-line no-undef
./platform/web/js/libs/library_godot_runtime.js:			out.apply(null, Array.from(arguments)); // eslint-disable-line no-undef
./platform/web/js/libs/library_godot_runtime.js:			return _malloc(p_size); // eslint-disable-line no-undef
./platform/web/js/libs/library_godot_runtime.js:			_free(p_ptr); // eslint-disable-line no-undef
./platform/web/js/libs/library_godot_runtime.js:			return getValue(p_ptr, p_type); // eslint-disable-line no-undef
./platform/web/js/libs/library_godot_runtime.js:			setValue(p_ptr, p_value, p_type); // eslint-disable-line no-undef
./platform/web/js/libs/library_godot_runtime.js:			return UTF8ToString(p_ptr); // eslint-disable-line no-undef
./platform/web/js/libs/library_godot_runtime.js:			return lengthBytesUTF8(p_str); // eslint-disable-line no-undef
./platform/web/js/libs/library_godot_runtime.js:			stringToUTF8(p_str, c_str, length); // eslint-disable-line no-undef
./platform/web/js/libs/library_godot_runtime.js:			return stringToUTF8Array(p_str, HEAP8, p_ptr, p_len); // eslint-disable-line no-undef
./platform/web/js/libs/library_godot_webgl2.js:		const gl_context_handle = _emscripten_webgl_get_current_context(); // eslint-disable-line no-undef
```

**Comment by patwork** [comment]
pluginJs.configs.all is crazy, with it's default set of rules eslint finds 4400+ problems:

```
   1366 camelcase
    479 no-magic-numbers
    467 func-names
    370 sort-keys
    301 object-shorthand
    277 one-var
    129 id-length
    127 prefer-arrow-callback
    118 capitalized-comments
     94 dot-notation
     71 func-style
     67 max-statements
     59 no-inline-comments
     49 no-unused-vars
     47 no-ternary
     47 max-params
     44 no-plusplus
     43 line-comment-position
     39 prefer-destructuring
     31 no-underscore-dangle
     29 no-undefined
     23 multiline-comment-style
     15 no-bitwise
     14 max-lines-per-function
     11 no-negated-condition
     11 max-lines
      7 no-self-assign
      6 no-implicit-coercion
      6 no-eq-null
      6 eqeqeq
      5 prefer-rest-params
      5 init-declarations
      4 require-unicode-regexp
      4 prefer-spread
      4 prefer-named-capture-group
      4 no-useless-assignment
      4 no-console
      4 consistent-this
      3 new-cap
      2 prefer-promise-reject-errors
      2 no-warning-comments
      2 no-empty-function
      1 no-continue
      1 max-classes-per-file
      1 func-name-matching
      1 complexity
```

**Comment by Faless** [comment]
> 1. I was wondering what is "onClientMessage" in service worker? I could not find documentation for such a function anywhere.

I actually think that's a bug (left-over dead code using undefined symbols).


> 2\. eslint upgrade will also require the removal of "eslint-disable" directives in several javascript files

Didn't we want to try to stay consistent with the rules? In that case most eslint-disable exceptions should be kept I think.
Happy to discuss which one requires changing.


> 3\. I was wondering whether, instead of ignoring errors in catch(), to add in each one something like console.warn(e)

Again, not sure where specifically, but some catch may be totally fine to fail, especially in engine code where we don't want to spam `WARN_PRINT`, and printing to console via emcc is actually not great in terms of performances.


> 4\. thanks to flat config, scripts in package.json can be simplified:

Looking great!

**Comment by patwork** [comment]
The thing is that config.all is more picky than the previous airbnb-base. After restoring all the old rules we're left with:

> 479 [no-magic-numbers](https://eslint.org/docs/latest/rules/no-magic-numbers)
> 370 [sort-keys](https://eslint.org/docs/latest/rules/sort-keys)
> 277 [one-var](https://eslint.org/docs/latest/rules/one-var)
> 129 [id-length](https://eslint.org/docs/latest/rules/id-length)
> 118 [capitalized-comments](https://eslint.org/docs/latest/rules/capitalized-comments)
>  59 [no-inline-comments](https://eslint.org/docs/latest/rules/no-inline-comments)
>  47 [no-ternary](https://eslint.org/docs/latest/rules/no-ternary)
>  47 [max-params](https://eslint.org/docs/latest/rules/max-params)
>  43 [line-comment-position](https://eslint.org/docs/latest/rules/line-comment-position)
>  29 [no-undefined](https://eslint.org/docs/latest/rules/no-undefined)
>  23 [multiline-comment-style](https://eslint.org/docs/latest/rules/multiline-comment-style)
>  11 [no-negated-condition](https://eslint.org/docs/latest/rules/no-negated-condition)
>   6 [no-implicit-coercion](https://eslint.org/docs/latest/rules/no-implicit-coercion)
>   6 [no-eq-null](https://eslint.org/docs/latest/rules/no-eq-null)
>   6 [eqeqeq](https://eslint.org/docs/latest/rules/eqeqeq)
>   5 [init-declarations](https://eslint.org/docs/latest/rules/init-declarations)
>   4 [require-unicode-regexp](https://eslint.org/docs/latest/rules/require-unicode-regexp)
>   4 [prefer-named-capture-group](https://eslint.org/docs/latest/rules/prefer-named-capture-group)
>   4 [no-useless-assignment](https://eslint.org/docs/latest/rules/no-useless-assignment)
>   4 [consistent-this](https://eslint.org/docs/latest/rules/consistent-this)
>   3 [new-cap](https://eslint.org/docs/latest/rules/new-cap)
>   2 [prefer-promise-reject-errors](https://eslint.org/docs/latest/rules/prefer-promise-reject-errors)
>   2 [no-warning-comments](https://eslint.org/docs/latest/rules/no-warning-comments)
>   2 [no-empty-function](https://eslint.org/docs/latest/rules/no-empty-function)
>   1 [func-name-matching](https://eslint.org/docs/latest/rules/func-name-matching)
>   1 [complexity](https://eslint.org/docs/latest/rules/complexity)

It's worth considering which ones are worth leaving enabled or disabling directly in the javascript code.

Also found two unnecessary eslint-disable:

> modules/webxr/native/library_godot_webxr.js
> 295:62   warning  Unused eslint-disable directive (no problems were reported from 'no-undef')

> platform/web/js/engine/features.js
> 1:20  warning  Unused eslint-disable directive (no problems were reported from 'no-unused-vars')

**Comment by patwork** [comment]
I think the first 13 rules that are repeated many times in the sources can be safely excluded. The others can be considered, as some of them may actually make it easier to avoid potential errors.

### Rule no-eq-null
```
/misc/dist/html/full-size.html
  153:10  error  Use '===' to compare with null  no-eq-null

/misc/dist/html/service-worker.js
   73:6  error  Use '===' to compare with null  no-eq-null
  125:9  error  Use '===' to compare with null  no-eq-null

/platform/web/js/engine/engine.js
   44:7  error  Use '===' to compare with null  no-eq-null
   81:9  error  Use '===' to compare with null  no-eq-null
  225:9  error  Use '===' to compare with null  no-eq-null

✖ 6 problems (6 errors, 0 warnings)
```

### Rule eqeqeq
```
/misc/dist/html/full-size.html
  153:23  error  Expected '!==' and instead saw '!='  eqeqeq

/misc/dist/html/service-worker.js
   73:15  error  Expected '===' and instead saw '=='  eqeqeq
  125:16  error  Expected '!==' and instead saw '!='  eqeqeq

/platform/web/js/engine/engine.js
   44:19  error  Expected '===' and instead saw '=='  eqeqeq
   81:21  error  Expected '===' and instead saw '=='  eqeqeq
  225:20  error  Expected '===' and instead saw '=='  eqeqeq

✖ 6 problems (6 errors, 0 warnings)
```

### Rule init-declarations
```
/misc/dist/html/editor.html
  421:5  error  Variable 'setStatusMode' should be initialized on declaration    init-declarations
  422:5  error  Variable 'setStatusNotice' should be initialized on declaration  init-declarations

/modules/webrtc/library_godot_webrtc.js
  52:9  error  Variable 'buffer' should be initialized on declaration  init-declarations

/modules/websocket/library_godot_websocket.js
  52:8  error  Variable 'buffer' should be initialized on declaration  init-declarations

/modules/webxr/native/library_godot_webxr.js
  417:7  error  Variable 'matrix' should be initialized on declaration  init-declarations

✖ 5 problems (5 errors, 0 warnings)
```

### Rule require-unicode-regexp
```
/platform/web/js/libs/library_godot_input.js
  243:17  error  Use the 'u' flag  require-unicode-regexp
  245:17  error  Use the 'u' flag  require-unicode-regexp
  419:25  error  Use the 'u' flag  require-unicode-regexp
  420:25  error  Use the 'u' flag  require-unicode-regexp

✖ 4 problems (4 errors, 0 warnings)
```

### Rule prefer-named-capture-group
```
/platform/web/js/libs/library_godot_input.js
  243:17  error  Capture group '([0-9a-f]{4})' should be converted to a named or non-capturing group  prefer-named-capture-group
  243:17  error  Capture group '([0-9a-f]{4})' should be converted to a named or non-capturing group  prefer-named-capture-group
  245:17  error  Capture group '([0-9a-f]+)' should be converted to a named or non-capturing group    prefer-named-capture-group
  245:17  error  Capture group '([0-9a-f]+)' should be converted to a named or non-capturing group    prefer-named-capture-group

✖ 4 problems (4 errors, 0 warnings)
```

### Rule no-useless-assignment
```
/modules/websocket/library_godot_websocket.js
  156:7  error  This assigned value is not used in subsequent statements  no-useless-assignment
  174:7  error  This assigned value is not used in subsequent statements  no-useless-assignment

/platform/web/js/engine/engine.js
  154:10  error  This assigned value is not used in subsequent statements  no-useless-assignment

/platform/web/js/libs/audio.worklet.js
  129:44  error  This assigned value is not used in subsequent statements  no-useless-assignment

✖ 4 problems (4 errors, 0 warnings)
```

### Rule consistent-this
```
/platform/web/js/engine/engine.js
   88:11  error  Unexpected alias 'me' for 'this'  consistent-this
  148:11  error  Unexpected alias 'me' for 'this'  consistent-this
  209:11  error  Unexpected alias 'me' for 'this'  consistent-this

/platform/web/js/engine/preloader.js
  111:10  error  Unexpected alias 'me' for 'this'  consistent-this

✖ 4 problems (4 errors, 0 warnings)
```

### Rule new-cap
```
/misc/dist/html/editor.html
  695:5  error  A function with a name starting with an uppercase letter should only be used as a constructor  new-cap

/platform/web/js/engine/engine.js
  96:8  error  A function with a name starting with an uppercase letter should only be used as a constructor  new-cap

/platform/web/js/libs/library_godot_runtime.js
  89:11  error  A function with a name starting with an uppercase letter should only be used as a constructor  new-cap

✖ 3 problems (3 errors, 0 warnings)
```

### Rule prefer-promise-reject-errors
```
/platform/web/js/libs/library_godot_input.js
  324:7  error  Expected the Promise rejection reason to be an Error  prefer-promise-reject-errors
  329:6  error  Expected the Promise rejection reason to be an Error  prefer-promise-reject-errors

✖ 2 problems (2 errors, 0 warnings)
```

### Rule no-warning-comments
```
/platform/web/js/libs/library_godot_os.js
  344:3  error  Unexpected 'todo' comment: 'TODO Godot core needs fixing to avoid...'  no-warning-comments

/platform/web/js/libs/library_godot_runtime.js
  94:62  error  Unexpected 'todo' comment: 'TODO wasm64'  no-warning-comments

✖ 2 problems (2 errors, 0 warnings)
```

### Rule no-empty-function
```
/modules/webxr/native/library_godot_webxr.js
  365:19  error  Unexpected empty arrow function  no-empty-function

/platform/web/js/libs/library_godot_os.js
  236:29  error  Unexpected empty method 'request_quit'  no-empty-function

✖ 2 problems (2 errors, 0 warnings)
```

### Rule func-name-matching
```
/misc/dist/html/editor.html
  676:3  error  Function name `progressFunction` should match property name `onProgress`  func-name-matching

✖ 1 problem (1 error, 0 warnings)
```

### Rule complexity
```
/modules/webxr/native/library_godot_webxr.js
  485:35  error  Method 'godot_webxr_update_input_source' has a complexity of 26. Maximum allowed is 20  complexity

✖ 1 problem (1 error, 0 warnings)
```

**Comment by patwork** [comment]
After further inspection:

### Rule no-eq-null

:question: Thoughts? Is it too strict?

### Rule eqeqeq

:question: Thoughts? Is it too strict?

### Rule init-declarations

:interrobang: Propably can be turned off, looks like it sometimes conflicts with _no-useless-assignment_.

### Rule require-unicode-regexp

:x: Off.

### Rule prefer-named-capture-group

:x: Off.

### Rule no-useless-assignment

:interrobang: Propably can be turned off, see _init-declarations_.

### Rule consistent-this

:question: Thoughts? Looks like it forces the use of the variable "that" when assigning "this".

> /*eslint consistent-this: ["error", "that"]*/

### Rule new-cap

:x: Off. Conflicts with "Godot", "UTF8ToString", etc.

### Rule prefer-promise-reject-errors

:question: Thoughts?

https://github.com/godotengine/godot/blob/2ba22d1554ded08094e3869a83bd87487266ad38/platform/web/js/libs/library_godot_input.js#L304-L332

### Rule no-warning-comments

:interrobang: Thoughts? These are user reminders, but will changing the level of this rule from "error" to "warn" result in failing github tasks?

### Rule no-empty-function

:x: Off.

### Rule func-name-matching

:question: Thoughts?

https://github.com/godotengine/godot/blob/2ba22d1554ded08094e3869a83bd87487266ad38/misc/dist/html/editor.html#L676

I don't think this function needs a name at all?

### Rule complexity

:x: Off.

**Comment by patwork** [comment]
@Faless I have set up a separate PR https://github.com/godotengine/godot/pull/91863

**Pull request closed without merging**